### PR TITLE
ENH: Reduce hyperparameters serialization time

### PR DESCRIPTION
**Pull request by kenza-bouzid**
Move encoder and pooling layer instantiation from the base container to deepmil lightning module.

### CODE REVIEW

File: hi-ml-histopathology/src/histopathology/utils/deepmil_params.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                                                 "keep the encoder frozen.")
+    is_caching: bool = param.Boolean(False, doc="If True, cache the encoded tile features "
+                                                "(disables random subsampling of tiles). "
+                                                "If False (default), load the tiles without caching "
+                                                "(enables random subsampling of tiles).")
+    encoding_chunk_size: int = param.Integer(0, doc="If > 0 performs encoding in chunks, by enconding_chunk_size tiles "
+                                                    "per chunk")
+
+
+class PoolingParams(param.Parameterized):
```
Comment (ant0nsc) [inline hi-ml-histopathology/src/histopathology/utils/deepmil_params.py:24]: add class docu please

File: hi-ml-histopathology/src/histopathology/models/deepmil.py [hunk 2]
```diff
@@ -34,81 +46,152 @@ class BaseDeepMILModule(LightningModule):
...
+            raise ValueError(f"Unsupported encoder type: {self.encoder_params.encoder_type}")
+
+        if self.encoder_params.is_finetune:
+            for params in encoder.parameters():
+                params.requires_grad = True
+        else:
+            encoder.eval()
+        return encoder
+
+    def get_pooling_layer(self, pooling_params: PoolingParams) -> Tuple[nn.Module, int]:
```
Comment (ant0nsc) [inline hi-ml-histopathology/src/histopathology/models/deepmil.py:164]: docu missing. Also, this function could live outside the class if I'm not mistaken, you'd only need `self.encoder.num_encoding` as an additonal arg?

File: hi-ml-histopathology/src/histopathology/models/deepmil.py [hunk 3]
```diff
@@ -34,81 +46,152 @@ class BaseDeepMILModule(LightningModule):
...
+            aml_workspace=get_workspace(),
+            run_id=ckpt_run_id,
+            checkpoint_filename=LAST_CHECKPOINT_FILE_NAME_WITH_SUFFIX,
+            download_dir=outputs_folder,
+            remote_checkpoint_dir=Path(f"{DEFAULT_AML_UPLOAD_DIR}/{CHECKPOINT_FOLDER}/")
+        )
+        downloader.download_checkpoint_if_necessary()
+        return downloader
+
+    def get_encoder(self, ckpt_run_id: Optional[str], outputs_folder: Optional[Path]) -> TileEncoder:
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-histopathology/src/histopathology/models/deepmil.py:120]: Good point, will do!

**Review by maxilse — APPROVED** [review]

File: hi-ml-histopathology/src/histopathology/utils/deepmil_utils.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    pool_hidden_dim: int = param.Integer(128, doc="If pooling has a learnable part, this defines the number of the\
+        hidden dimensions.")
+    pool_out_dim: int = param.Integer(1, doc="Dimension of the pooled representation.")
+    num_transformer_pool_layers: int = param.Integer(4, doc="If transformer pooling is chosen, this defines the number\
+         of encoding layers.")
+    num_transformer_pool_heads: int = param.Integer(4, doc="If transformer pooling is chosen, this defines the number\
+         of attention heads.")
+
+
+def get_encoder(
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-histopathology/src/histopathology/utils/deepmil_utils.py:52]: I like your suggestion, will put it in place.

File: hi-ml-histopathology/testhisto/testhisto/models/test_deepmil.py [hunk 5]
```diff
@@ -35,20 +35,18 @@
 no_gpu = not is_gpu_available()
 
 
-def get_supervised_imagenet_encoder() -> TileEncoder:
-    return ImageNetEncoder(feature_extraction_model=resnet18, tile_size=224)
+def get_supervised_imagenet_encoder_params() -> EncoderParams:
+    encoder_params = EncoderParams()
+    encoder_params.encoder_type = ImageNetEncoder.__name__
```
Comment (dccastro) [inline hi-ml-histopathology/testhisto/testhisto/models/test_deepmil.py:40]: You should be able to instantiate `param` classes using constructor keyword arguments, no?

**Pull request merged**